oeps/oep-0006.rst

```diff
@@ -0,0 +1,224 @@
+
+
+Rationale
+=========
+
+Many of the ideas in this proposal are driven by the desire for new types of functionality to be implemented via XBlocks, and thus to be pluggable and easily installed, enabled, or disabled for each Open edX installation or each course.
+
+The idea of naming the new scope "context" comes from other technologies such as LTI, where the term "context" is used analogously. For example, when embedding an LTI tool into the Open edX LMS, the LMS will pass the current course ID as the LTI "context" parameter.
+
+Previous discussions of note include:
+
+* https://openedx.atlassian.net/browse/PLAT-325
+* https://groups.google.com/d/msg/edx-code/ywjXV0wzQiw/FRzaK5nTAgAJ
+* On 2016-07-08, the author met with several stakeholders from edX (Calen Pennington, David Ormsbee, and Robert Raposa) to discuss course-scoped XBlock fields. The discussion tended toward using "context" as an additional dimension (see "Context as a New Scope Dimension" in "Rejected Alternatives"), but we did not think of any use cases for 8 of the 12 additional scopes that would be created by defining an additional dimension.
+
+Other related proposals of note include:
+
+* https://github.com/edx/XBlock/pull/317 - XBlock field sharing
+
+
+Backward Compatibility
+=======================
+
+This OEP introduces a new field scope but does not remove or modify any existing functionality. XBlock authors that add new fields to their XBlock may opt-in to using this feature.
+
+XBlock runtimes may choose to implement context-scoped fields or not. Attempting to access a context-scoped field in a runtime that does not support such fields should simply return the default value of the field. Attempting to modify such a context-scoped field in a runtime that doesn't support such fields should raise an ``InvalidScopeError``.
+
+Some existing XBlocks currently use workarounds for the lack of context-scoped fields, such as storing data in the fields of the root 'course' XBlock. Converting such existing techniques to use context-scoped fields instead would be best achieved with an XBlock data migration API, which currently does not exist and is outside the scope of this OEP (however, see `here`_ for an example implementation).
+
+.. _here: https://github.com/open-craft/problem-builder/blob/c5879dd/problem_builder/v1/xml_changes.py
+
+Details on the current implementation of field scopes in edx-platform can be found in `this gist`_. The `field test XBlock`_ can be used to test how a runtime behaves with different field scopes.
+
+.. _this GIST: https://gist.github.com/bradenmacdonald/b576c59e46af8df9949fa84b3bc7a4b6
+.. _field test XBlock: https://github.com/open-craft/xblock-field-test
+
+Open Questions
+==============
+
+#. How can an XBlock specify whether a context-scoped field should participate
```

Comment:
@andy-armstrong That sounds good to me. If we go that route (each XBlock can declare a view for its context-scoped fields), I'd like to make sure we standardize on something like [https://github.com/edx/xblock-utils/#studioeditablexblockmixin](https://github.com/edx/xblock-utils/#studioeditablexblockmixin) so that basic boilerplate UI can easily be generated and we don't get XBlock authors coding idiosyncratic settings UIs where they aren't needed.

There's also the question of how to edit fields with scope `Scope.context_shared_settings` (course-wide settings shared among multiple XBlock types), but I guess the answer there is that developers would have to choose one XBlock to implement the UI, or allow setting the UI via the course-wide settings view of multiple XBlock types.

Oh, that also gets close to the "course_view" concept that we'd also [discussed on the mailing list](https://groups.google.com/d/msg/edx-code/ywjXV0wzQiw/uBeIkoyjBAAJ), which is something that this work is intended to enable.